### DISCUSSION TITLE
Dictionary Helper

### DISCUSSION THREAD

**Discussion by dreamtheater39** [home-assistant/architecture#782] [Entity Models] [answered]
Proposing a Dictionary Helper to be added to HASS along with input_select and input_text helpers. The idea of this is to add custom key/value pairs at run time that can used as variable for automations/scripts.

I wasn't aware of an approval process and have actually gone ahead and built a component that does this. The component currently supports all necessary functions - save/restore, history etc. but is missing a frontend implementation which i plan to do next (hopefully with some community support!)

Currently - you can create dictionary entities through configuration.yaml and add/remove/edit keys at runtime. You can query the state of these keys to use in automations and scripts.

The motivation to create this component was to have a more expressive helper while creating automations. Many folks use the input_select helper to get around this - and end up using it in ways it wasn't designed!
 
Here's the Github url for the current component
https://github.com/dreamtheater39/input_dictionary

Here's the HA/core fork where it's integrated as one of the built in components
https://github.com/dreamtheater39/core/tree/input_dictionary

**Comment by frenck** [comment]
We have recently support allowing dictionaries as options to the select selector.
I think it is a good idea to add those to other places as well.

However, I would not add it to `input_select`, but to the `select` base entity component.

I'm not sure what the input text helper has to do with his?

**Reply by dreamtheater39** [reply]
I only meant a Dictionary helper can be an additional helper (alongside text,numeric,input_select helpers).

You could test out the component on my git link - and also look at the examples below to get a better idea -
https://github.com/dreamtheater39/input_dictionary

Also, i've created a new component input_dictionary

**Reply by frenck** [reply]
> I only meant a Dictionary helper can be an additional helper 

We don't need helpers for that, that is already natively supported by using variables & jinja.

But having key/value pair in select entities (and thus also `input_select`), makes sense to me.

**Reply by dreamtheater39** [reply]
input_select is being used as drop-down pickers etc in dashboards. From a ux standpoint - it may make things confusing for a non savvy user don't you think?

So it may make more sense to separate the two functionalities and not disturb Drop-down helper as it exists today?

**Reply by frenck** [reply]
> From a ux standpoint - it may make things confusing for a non savvy user don't you think?

Could you give an illustrated example of what is wrong? That statement by itself doesn't show the problem.

**Reply by dreamtheater39** [reply]
For starters you'd have to rename Dropdown to be something more reflective of the new capabilities that a dictionary would bring. A dictionary is more like a "register" in human language.

The functionality of Dropdown helper will conflict with Dictionary capabilities. The component I've built right now allows for quick store/append/retrieve capability that you can use....and can be an essential part of the toolkit while building automations.

One option is to merge implementation of dictionary with input_select and have 2 user facing helpers

**Reply by frenck** [reply]
> For starters you'd have to rename Dropdown to be something more reflective of the new capabilities that a dictionary would bring.

No? The select entity could just accept both in their schema.

>  A dictionary is more like a "register" in human language.

That is exactly what the mapping in options are for. e.g., this selector example:

```yaml
select:
  options:
    - label: Red
      value: r
    - label: Green
      value: g
    - label: Blue
      value: b
```

See: <https://www.home-assistant.io/docs/blueprint/selectors/#select-selector>


> The functionality of Dropdown helper will conflict with Dictionary capabilities.

Sorry, I don't see that issue.

> and can be an essential part of the toolkit while building automations.

Agree, that would make live easier and is IMHO simply a change of allowing key/value dictionaries on selects, just like we did in the selectors recently.

**Reply by dreamtheater39** [reply]
Sure, that can be done too.....but the front facing name of "Dropdown" should change to something else? And an input select exposes services like "select next, previous, first etc." Which may not make much sense if you're trying to use the entity as a dictionary

**Reply by frenck** [reply]
> Sure, that can be done too.....but the front facing name of "Dropdown" should change to something else

It does, I'm not following your train of thought here...

>  "select next, previous, first etc." Which may not make much sense if you're trying to use the entity as a dictionary

Why not? Both are sets of items that can be iterated on.

**Reply by dreamtheater39** [reply]
But use cases are different from a human understandable form. Developers or nerds would figure how to use something that's called "Dropdown" as a dictionary.... But normal folks wouldnt!

Did you take a look at my git url? See the use cases there.....i don't see how that'll fit in with something thats called "Dropdown"

**Reply by frenck** [reply]
> But use cases are different from a human understandable form. Developers or nerds would figure how to use something that's called "Dropdown" as a dictionary.... But normal folks wouldn't!

I think we are talking about the same things, but you are missing the alternative suggestion I try to make. Allowing dictionary key/value options on select entities (which includes input_select), means the internal value for developers can be different than the value shown to the end user.

> Did you take a look at my git url? See the use cases there....

I did. Most of the things happening there can be done with existing Jinja templating and variables and don't need a custom integration.

> .i don't see how that'll fit in with something thats called "Dropdown"

A drop-down isn't a single value. It can have a "display" value and a "return" value paired in a dictionary. See the linked example on the select selector above.

For example, the Dropdown can show "Turn on all lights", while the value returned is "all_on". 
We can decouple the value shown vs value returned as done in our select selector already.

**Reply by dreamtheater39** [reply]
Sure - let me think about it for a day :)
Also, Select entities can only be created in configuration.yaml right? There isnt a frontend helper to create your own Select entities if i'm not wrong?

**Reply by frenck** [reply]
> Also, Select entities can only be created in configuration.yaml right?

No `select` entities are only provided by other integrations. A user can create `input_select` entities (which child of `select` so it inherits its features). The `input_select` is also available from the UI (Dropdown helper, so configurable via both YAML and UI).

**Reply by dreamtheater39** [reply]
Let me check if i got this right....
1. Upgrade input_select helper to accept values to each of the drop-down options (work on frontend mainly)
2. Implementation of this will reside in the Select entity code....so all other entities that inherit will benefit
3. Storage/Retrieval of values will be done in Jinja variables? Not sure how we can do this now....without creating either a number or text or Dropdown helper

**Comment by emontnemery** [comment, marked as answer]
This proposal was discussed with the core team July 3rd, and we don't want to go forward with it at this time
